Environment variables in config

alexoneill opened the pull request:
This adds support for environment variables in config files.

This feature is primarily intended for obfuscating sensitive information within configuration files, such as usernames, passwords, API keys, etc. That way, you can share your config without spoiling secrets!

You can take a look at the documentation I added below, but essentially the changes enable the following functionality:

```python
order += "my_module"
order += env(ORDER_MODULE)

module {
    normal_parameter = 'some value'
    env_parameter = env(SOME_ENVIRONMENT_PARAM)
    sensitive_api_key = env(API_KEY)

    complex_parameter = {
      'key': env(VAL)
    }

    list_of_tuples = [
      (env(APPLE_NUM, int), 'apple'),
      (2, env(ORANGE))
    ]

    float_param = env(MY_NUM, float)
}
```

In total, this adds:

- Environment variable support in configuration files via `env(...)`
- Validation of environment variables to ensure presence in the system
- Support for converting values in the variables to specific data types

As always, let me know if there is anything I should update or change! :)

alexoneill commented on py3status/parse_config.py:
```diff
@@ -632,19 +685,23 @@ def remove_any_contained_modules(module):
                 fixed[k] = v
         return fixed
 
+    def append_modules(item):
```
This was just a little bit of code cleanup I noticed :)

ultrabug replied:
okay for this time, but please separate these kind of PRs next time

alexoneill replied:
I removed it and moved it to a new PR! :)

ultrabug commented:
Nice, I love it thanks a lot @alexoneill !

Will check it out ASAP

tobes commented:
I also think this is a nice feature.  My thoughts are

A. Should we auto guess the type. We do this already I believe. 

B. We could just do `$VARIABLE` this sort of needs the above type guessing

alexoneill commented:
@tobes I like the idea, but what if the user wants to have a string that's `'12345'`? Would we correct that to the number form with the type guessing?

tobes commented:
> but what if the user wants to have a string that's '12345'

In the config if a value is quoted then we treat it as a string.  Anything else we try to type guess as a boolean / float / int / None.

As it is defining types in the shell is quite hard/error prone but I'm not sure if this causes us any real problems.

```shell
$ export moo='12345'
$ echo $moo
12345
$ export moo=12345
$ echo $moo
12345
$ export moo=\'12345\'
$ echo $moo
'12345'
```

sorry for the late reply

alexoneill commented:
@tobes No worries! :)

I changed the feature to use `$...` as an environment variable identifier. Plus, I added that type guessing you mentioned!

@ultrabug I changed the docs to reflect that this is new in v3.8!

Alex

lasers commented:
I wanted to share that I like `env(EXAMPLE)` more than `$EXAMPLE`.
1) because it appears to be more pythonic... like how python interpreters uses `exit()` to exit.. and shells uses `exit`.
2) this could make noises on python syntax.
3) seeing `env()` would be more obvious telling that it is using `environment variables` than just `$` and maybe feels more clean too.

```python
# PYTHON
order += "my_module"
order += env(ORDER_MODULE)

module {
    normal_parameter = 'some value'
    env_parameter = env(SOME_ENVIRONMENT_PARAM)
    sensitive_api_key = env(API_KEY)

    complex_parameter = {
      'key': env(VAL)
    }

    list_of_tuples = [
      (env(APPLE_NUM, int), 'apple'),
      (2, env(ORANGE))
    ]

    float_param = env(MY_NUM, float)
}
```
```python
# SHELL
order += "my_module"
order += $ORDER_MODULE

module {
    normal_parameter = 'some value'
    env_parameter = $SOME_ENVIRONMENT_PARAM
    sensitive_api_key = $API_KEY

    complex_parameter = {
      'key': $VAL
    }

    list_of_tuples = [
      ($APPLE_NUM, int), 'apple'),  # probably wrong, auto-guessing.
      (2, $ORANGE)
    ]
    float_param = $MY_NUM, float    # probably wrong, auto-guessing.
}
```

alexoneill commented:
@lasers I agree, and it's easy to roll it back to that syntax. To stay 'pythonic' do you think the environment variables should be surrounded in quotes?

lasers commented:
```python
# PYTHON w/ QUOTES
order += "my_module"
order += env('ORDER_MODULE')

module {
    normal_parameter = 'some value'
    env_parameter = env('SOME_ENVIRONMENT_PARAM')
    sensitive_api_key = env('API_KEY')

    complex_parameter = {
      'key': env('VAL')
    }

    list_of_tuples = [
      (env('APPLE_NUM', int), 'apple'),
      (2, env('ORANGE'))
    ]

    float_param = env('MY_NUM', float)
}
```
I have no opinion on this. Is it trivial to support both? I think we should leave it as-is... This way, they all would look like CONSTANT instead of 'strings'.

What I think would be cool is if we can support this out of box without all this in the config...
```python
general {
    color1 = env(COLOR1)
    color2 = env(COLOR2)
    color3 = env(COLOR3)
    color4 = env(COLOR4)
    color5 = env(COLOR5)
    color6 = env(COLOR6)
    color7 = env(COLOR7)
    color8 = env(COLOR8)
    color9 = env(COLOR9)
    color10 = env(COLOR10)
    color11 = env(COLOR11)
    color12 = env(COLOR11)
    color13 = env(COLOR13)
    color14 = env(COLOR14)
    color15 = env(COLOR15)
    color16 = env(COLOR16)
}
```
then users can incoporate `py3status` in their color/theme/others script changers... eg setting a wallpaper plus getting 16 colors from wallpaper to configure GTK themes, i3 windows, i3bar colors, `py3status` colors, etc.

I'm not sure what to do with `good`, `degraded`, and `bad`. I think we add this too...
```python
general {
    color_bad = env(COLOR1)
    color_degraded = env(COLOR2)
    color_good = env(COLOR3)
}
```
This way, we don't have to change anything and `py3status` would have same color as `terminal` (or others). If no colors, then we assume normal `good`, `degraded`, `bad`.... for this one... Maybe all in next PR or something for `3.8`. :-)

alexoneill commented:
@lasers @tobes 

I reverted back to `env(...)` notation, but I kept type guessing and made it the default conversion mode. That way there is on average less configuration, as well as the ability to force a certain type conversion if need be.

I like the idea of colors populated from environment variables, but to keep this from getting too large like some of my _cough cough_ other pull requests, I'll break it out into a separate PR. :)

Alex

lasers commented:
>I like the idea of colors populated from environment variables.

I wonder if it's better to parse colors from `xrdb -query`?

tobes commented:
> I reverted back to env(...) notation, 

:( To me `$VAR` is obvious where as `env(VAR)` looks weird in a config file

> I wonder if it's better to parse colors from xrdb -query?

This is an interesting idea but outside of this PR.  I should revive my named colours branch for 3.8

girst commented:
Please excuse my interjection,

girst commented:
Please excuse my interjection, but I'd prefer `env(...)` as it makes clear where the variable is coming from. `$VAR` could as well be a variable internal to py3staus' config.

tobes commented:
> Please excuse my interjection

feedback and opinion is always welcome, thanks for these thoughts

ultrabug commented:
Then let's settle for env() thanks a lot.

@girst this is perfect as community voting is what we're looking for!

@tobes let's do a one time code review together tomorrow ;)

ultrabug commented on py3status/parse_config.py:
```diff
@@ -292,6 +310,41 @@ def make_value(self, value):
 
+    def make_value_from_env(self, value):
+        # Extract the environment variable and type
+        match = re.match('env\(([0-9A-Z_]+)(, *[a-z]+)?\)', value)
```
the `(, *[a-z]+)` should be `(,\s?*[a-z]+)` because there's a high chance that they forget the space

girst replied:
erm; ` *` matches zero-or-more spaces; `\s?*` at least to me looks like a syntax error (and the [docs](https://docs.python.org/3/library/re.html#regular-expression-syntax) seem to think so as well)

tobes replied:
`'env\(\s*([0-9A-Z_]+)(,\s*[a-z]+)?\s*\)'`

Let's just not be strict with the whitespace.  Also we are being case sensitive is there a reason we are doing this?

ultrabug replied:
You're right @girst I was sloppy and missed the *

I agree with @tobes proposal

alexoneill replied:
No reason for the lowercase, I threw a `string.lower()` call in there to be compatible. Thanks for the updated regex! I added support for mixed case and underscores as well.

ultrabug commented on py3status/parse_config.py:
```diff
@@ -292,6 +310,41 @@ def make_value(self, value):
+        # TODO: dynamic support, (lambda: os.getenv(env_var))
+        value = os.getenv(env_var)
+        if value is None:
+            self.error('Environment variable undefined')
```
please add the name of the env var in the error message

alexoneill replied:
Fixed!

ultrabug commented on py3status/parse_config.py:
```diff
@@ -292,6 +310,41 @@ def make_value(self, value):
+            'auto': self.make_value,
+        }
+        if env_type not in conversion_options:
+            self.error('Invalid conversion')
```
I think this message and the one below (bad conversion) should be more explicit like : "invalid/bad env conversion to env_type"

alexoneill replied:
Fixed!

tobes commented on py3status/parse_config.py:
```diff
@@ -292,6 +311,42 @@ def make_value(self, value):
 
+    def make_value_from_env(self, value):
+        # Extract the environment variable and type
+        match = re.match('env\(\s*([0-9A-Z_]+)(\s*,\s*[a-zA-Z_]+)?\s*\)', value)
```
Why not allow vars that are lower case?

'env\\(\s*([0-9**a-z**A-Z_]+)(\s*,\s*[a-zA-Z_]+)?\s*\\)'

alexoneill replied:
Sure! Fixed. :)

ultrabug commented:
Thanks a lot everyone. @tobes I think this is ready now.

tobes commented:
@alexoneill could you resolve the merge conflict and rebase this then I think we are good to go.

Thanks for all you efforts

ultrabug commented:
Awesome feature @alexoneill thanks a lot!
